src/cargo/core/registry.rs

```diff
@@ -143,20 +143,23 @@ impl<'cfg> PackageRegistry<'cfg> {
             }
         }
 
-        try!(self.load(namespace, Kind::Normal));
+        let mut updated = HashSet::new();
+        try!(self.load(namespace, Kind::Normal, &mut updated));
         Ok(())
     }
 
     pub fn add_sources(&mut self, ids: &[SourceId]) -> CargoResult<()> {
+        let mut updated = HashSet::new();
         for id in ids.iter() {
-            try!(self.load(id, Kind::Locked));
+            try!(self.load(id, Kind::Locked, &mut updated));
```

Comment:
You may want to try using `ensure_loaded` here instead of `load` as that should handle the deduplication for you